### PR TITLE
Switch Windows Checked build to `-O2` from `-O1`

### PR DESCRIPTION
**Pull request by BruceForstall**
That is, optimize for speed, not for size.

Related: https://github.com/dotnet/runtime/issues/53849

### CODE REVIEW

File: eng/native/configureoptimization.cmake [hunk 1]
```diff
@@ -1,6 +1,6 @@
 if(CLR_CMAKE_HOST_WIN32)
     add_compile_options($<$<AND:$<COMPILE_LANGUAGE:C,CXX>,$<CONFIG:Debug>>:/Od>)
-    add_compile_options($<$<AND:$<COMPILE_LANGUAGE:C,CXX>,$<CONFIG:Checked>>:/O1>)
+    add_compile_options($<$<AND:$<COMPILE_LANGUAGE:C,CXX>,$<CONFIG:Checked>>:/O2>)
     add_compile_options($<$<AND:$<COMPILE_LANGUAGE:C,CXX>,$<CONFIG:Release>>:/Ox>)
```
Comment (tannergooding) [inline eng/native/configureoptimization.cmake:4]: Should this also be `/O2`, it seems odd that Checked and RelWithDebInfo are "more optimized" than `Release`.
-- More optimized in that `O2` is (`Ox` + `GF` + Gy`)

Comment (tannergooding) [reply eng/native/configureoptimization.cmake:4]: and maybe doesn't include `Og` although it isn't explicitly called out on https://docs.microsoft.com/en-us/cpp/build/reference/ox-full-optimization?redirectedfrom=MSDN&view=msvc-160, so its just an observed difference from https://docs.microsoft.com/en-us/cpp/build/reference/o1-o2-minimize-size-maximize-speed?redirectedfrom=MSDN&view=msvc-160

**Comment by EgorBo** [comment]
So judging by https://github.com/dotnet/runtime/issues/53849#issuecomment-856969889
it adds +1.7Mb for all apps, even for a hello world?

**Comment by BruceForstall** [comment]
> it adds +1.7Mb for all self-contained apps, even for a hello world?

I didn't measure any "user" SDK-style metrics, since I never do that (with a Checked build).

**Comment by EgorBo** [comment]
> > it adds +1.7Mb for all self-contained apps, even for a hello world?
> 
> I didn't measure any "user" SDK-style metrics, since I never do that (with a Checked build).

Ah, I didn't realize it's Checked only, then it's nice (TP improvements), I hope it doesn't increase compilation time 🙂

**Comment by BruceForstall** [comment]
Well, this being the CLR, the VM is very sensitive to native compiler optimization, so there are x86 test failures, e.g.,
```
Assert failure(PID 7704 [0x00001e18], Thread: 7708 [0x1e1c]): FastNExportSEH(pEHR) || IsComPlusNestedExceptionRecord(pEHR) || ((ExecuteHandler2Module != NULL) && IsIPInModule(ExecuteHandler2Module, (PCODE)pEHR->Handler))

CORECLR! PopNestedExceptionRecords + 0xA0 (0x71127c40)
CORECLR! COMPlusEndCatch + 0x4A (0x711285da)
CORECLR! JIT_EndCatch + 0xE (0x71199e41)
CORECLR! CallDescrWorkerInternal + 0x34 (0x71199121)
CORECLR! CallDescrWorker + 0x13E (0x70da52ee)
CORECLR! CallDescrWorkerWithHandler + 0x189 (0x70da5589)
CORECLR! MethodDescCallSite::CallTargetWorker + 0x7A7 (0x70da5d77)
CORECLR! RunMainInternal + 0x2FE (0x70ae232e)
CORECLR! ``RunMain'::29'::__Body::Run'::5'::__Body::Run + 0x86 (0x70ae1c36)
CORECLR! RunMain'::29'::__Body::Run + 0xA3 (0x70ae1d53)
File: D:\workspace_work\1\s\src\coreclr\vm\i386\excepx86.cpp Line: 2054
Image: C:\h\w\B10F09DA\p\corerun.exe
```
and:
```
Assert failure(PID 3932 [0x00000f5c], Thread: 5880 [0x16f8]): Consistency check failed: Invalid transition into managed code!

We're walking this thread's stack and we've reached a managed frame at Esp=0x00F7D338. (The method is EHTest::f1) The very next FS:0 record (0x00F7D460) up from this point on the stack should be one of our 'unmanaged to managed SEH handlers', but its not... its something else, and that's very bad. It indicates that someone managed to call into managed code without setting up the proper exception handling.

Get a good unmanaged stack trace for this thread. All FS:0 records are on the stack, so you can see who installed the last handler. Somewhere between that function and where the thread is now is where the bad transition occurred.

A little extra info: FS:0 = 0x00F7C35C, pEHR->Handler = 0x70D12C3A
FAILED: IsUnmanagedToManagedSEHHandler(pEHR)

CORECLR! CHECK::Trigger + 0x3C4 (0x70a75264)
CORECLR! VerifyValidTransitionFromManagedCode + 0x18A (0x709e856a)
CORECLR! StackFrameIterator::NextRaw + 0xAE4 (0x705ca434)
CORECLR! Thread::StackWalkFramesEx + 0x421 (0x705cb681)
CORECLR! Thread::StackWalkFrames + 0x159 (0x705cb1c9)
CORECLR! LookForHandler + 0x1A (0x704959da)
CORECLR! CPFH_RealFirstPassHandler + 0xF57 (0x709e6007)
CORECLR! CPFH_FirstPassHandler + 0x1C0 (0x709e5020)
CORECLR! COMPlusFrameHandler + 0x26B (0x709e88db)
CORECLR! COMPlusNestedExceptionHandler + 0x13B (0x709e8a6b)
File: D:\workspace_work\1\s\src\coreclr\vm\i386\excepx86.cpp Line: 329
Image: C:\h\w\B10F09DA\p\corerun.exe
```

**Comment by janvorli** [comment]
It looks like one of the the infamous unwinding issues on x86, compiler generating instructions that the code in src/coreclr/src/vm/i386/gmsx86.cpp cannot unwind correctly.

**Pull request closed without merging**